sycl/unittests/Extensions/DeviceGlobal.cpp

```diff
@@ -215,3 +302,569 @@ TEST(DeviceGlobalTest, DeviceGlobalInitBeforeUse) {
+      after_piextUSMEnqueueMemset);
+  Mock.redefineAfter<
+      sycl::detail::PiApiKind::piextEnqueueDeviceGlobalVariableWrite>(
+      after_piextEnqueueDeviceGlobalVariableWrite<true>);
+  Mock.redefineAfter<sycl::detail::PiApiKind::piEventGetInfo>(
+      after_piEventGetInfo);
+
+  // Create new context to isolate device_global initialization.
+  sycl::context C{Plt.get_devices()[0]};
+  sycl::queue Q{C, sycl::default_selector_v};
+
+  int Vals[2] = {42, 1234};
+  Q.copy(Vals, DeviceGlobal).wait();
+
+  // Device global should not have been written to yet. Fill operation is not
+  // needed since the device_global will have a new value.
+  EXPECT_TRUE(!DeviceGlobalWriteEvent.has_value());
+  EXPECT_TRUE(!DeviceGlobalFillEvent.has_value());
+
+  Q.single_task<DeviceGlobalTestKernel>([]() {});
+
+  // The device global should now have its USM memory pointer written, but fill
+  // should still not have happened as an explicit write have happened to the
+  // underlying memory.
+  EXPECT_TRUE(DeviceGlobalWriteEvent.has_value());
+  EXPECT_TRUE(!DeviceGlobalFillEvent.has_value());
+}
+
+TEST(DeviceGlobalTest, DeviceGlobalMemcpyToBeforeUseFull) {
+  ResetTrackersAndMarkers();
+
+  sycl::unittest::PiMock Mock;
+  sycl::platform Plt = Mock.getPlatform();
+
+  Mock.redefineAfter<sycl::detail::PiApiKind::piextUSMEnqueueMemset>(
+      after_piextUSMEnqueueMemset);
+  Mock.redefineAfter<
+      sycl::detail::PiApiKind::piextEnqueueDeviceGlobalVariableWrite>(
+      after_piextEnqueueDeviceGlobalVariableWrite<true>);
+  Mock.redefineAfter<sycl::detail::PiApiKind::piEventGetInfo>(
+      after_piEventGetInfo);
+
+  // Create new context to isolate device_global initialization.
+  sycl::context C{Plt.get_devices()[0]};
+  sycl::queue Q{C, sycl::default_selector_v};
+
+  int Vals[2] = {42, 1234};
+  Q.memcpy(DeviceGlobal, Vals).wait();
+
+  // Device global should not have been written to yet. Fill operation is not
+  // needed since the device_global will have a new value.
+  EXPECT_TRUE(!DeviceGlobalWriteEvent.has_value());
+  EXPECT_TRUE(!DeviceGlobalFillEvent.has_value());
+
+  Q.single_task<DeviceGlobalTestKernel>([]() {});
+
+  // The device global should now have its USM memory pointer written, but fill
+  // should still not have happened as an explicit write have happened to the
+  // underlying memory.
+  EXPECT_TRUE(DeviceGlobalWriteEvent.has_value());
+  EXPECT_TRUE(!DeviceGlobalFillEvent.has_value());
+}
+
+TEST(DeviceGlobalTest, DeviceGlobalCopyToBeforeUsePartialNoOffset) {
+  ResetTrackersAndMarkers();
+
+  sycl::unittest::PiMock Mock;
+  sycl::platform Plt = Mock.getPlatform();
+
+  Mock.redefineAfter<sycl::detail::PiApiKind::piextUSMEnqueueMemset>(
+      after_piextUSMEnqueueMemset);
+  Mock.redefineAfter<
+      sycl::detail::PiApiKind::piextEnqueueDeviceGlobalVariableWrite>(
+      after_piextEnqueueDeviceGlobalVariableWrite<true>);
+  Mock.redefineAfter<sycl::detail::PiApiKind::piEventGetInfo>(
+      after_piEventGetInfo);
```

Comment:
These are repeated in several testcase, I think we should move the common part into a class.

Reply:
There is actually no common redefinitions. Some may be possible to have for all, but some make assumptions that will not work for all tests.

Reply:
Still, I'd like to hide (at least)

```
  ResetTrackersAndMarkers();

  sycl::unittest::PiMock Mock;
  sycl::platform Plt = Mock.getPlatform();
```

and (if possible)

```
  // Create new context to isolate device_global initialization.
  sycl::context C{Plt.get_devices()[0]};
  sycl::queue Q{C, sycl::default_selector_v};
```

somehow...

Reply:
I've added a `CommonSetup` function that takes a function for doing redefinitions inbetween the two blocks you mention. Is that in line with what you had in mind?